### ISSUE TITLE
Error บน wongnai_reviews

### ISSUE THREAD

**Issue by peune** [vistec-AI/thai2transformers#65]
สวัสดีครับ 

ผมลองใช้ wangchanberta ดูบน wongnai_reviews ตามโค้ดข้างล่างนี้ แล้วเจอ error แปลก ๆ ไม่ทราบว่าต้องแก้อย่างไรครับ
```
from transformers import (
    CamembertTokenizer,
    AutoModelForSequenceClassification,
    pipeline
)
from thai2transformers.preprocess import process_transformers

# Load pre-trained tokenizer
tokenizer = CamembertTokenizer.from_pretrained(
                                  'airesearch/wangchanberta-base-att-spm-uncased',
                                  revision='main')
tokenizer.additional_special_tokens = ['<s>NOTUSED', '</s>NOTUSED', '<_>']

# Load pre-trained model
model = AutoModelForSequenceClassification.from_pretrained(
                                  'airesearch/wangchanberta-base-att-spm-uncased',
                                  revision='finetuned@wongnai_reviews')

classify_sequence = pipeline(task='text-classification',
          tokenizer=tokenizer,
          model=model)

from datasets import load_dataset

dataset = load_dataset('wongnai_reviews', cache_dir='huggingface_cache')

text = dataset['test'][3]['review_body']

processed_input_text = process_transformers(text)

result = classify_sequence(processed_input_text)
print(result)
```

Error ที่เจอคือ
![image](https://user-images.githubusercontent.com/13883651/129322181-4d386e3c-02d1-4a01-837a-cdd63feba756.png)

**Comment by artificiala** [comment]
สวัสดีค่ะ คุณ @peune,

กรณี RuntimeError ดังกล่าว เกิดขึ้นเพราะ จำนวน tokens (รวม special tokens เช่น \<s\>, \<s\/\>) มีจำนวนเกิดน 510 tokens (ซึ่งเป็นข้อจำนวนของ architecture ประเภท transformer encoder)

สำหรับวิธีการแก้ไขโดยเบื้องต้นคือการทำ text truncation ให้ sequence ที่ส่งเข้าไปที่โมเดลมีจำนวนไม่เกินที่กำหนด โดยการระบุ max_length

```python
text = dataset['test'][3]['review_body']

processed_input_text = process_transformers(text)

# Truncate text
processed_input_text_truncated = ''.join(tokenizer.tokenize(processed_input_text)[0:508]).replace('▁', '')

result = classify_sequence(processed_input_text_truncated)
print(result)
```

Output

```
[{'label': '4', 'score': 0.5953860878944397}]
```

ตัวอย่างใน Google Colab: https://colab.research.google.com/drive/1MA6fJdbYzk_G2SZ7czD6pCEY6C3omgDI?usp=sharing

**Comment by peune** [comment]
ขอบคุณครับ

ขอถามต่ออีกนิดครับ ผมลองเล่น wangchanberta กับ wongnai_reviews ต่อ แต่ได้ accuracy ประมาณ 16.9% เองครับ ไม่ทราบว่าทำอะไรผิดตรงไหน เห็นบนเว็บบอกว่าได้ 63% กว่าๆ

ผมใช้โค้ดตามข้างล่างนี้ครับ


```
#thai2transformers
import thai2transformers
from thai2transformers.preprocess import process_transformers
from thai2transformers.metrics import (
    classification_metrics, 
    multilabel_classification_metrics,
)
from thai2transformers.tokenizers import (
    ThaiRobertaTokenizer,
    ThaiWordsNewmmTokenizer,
    ThaiWordsSyllableTokenizer,
    FakeSefrCutTokenizer,
    SEFR_SPLIT_TOKEN
)

from transformers import (
    CamembertTokenizer,
    AutoTokenizer,
    AutoModel,
    AutoModelForMaskedLM,
    AutoModelForSequenceClassification,
    AutoModelForTokenClassification,
    TrainingArguments,
    Trainer,
    pipeline,
)

from transformers import (
    CamembertTokenizer,
    AutoModelForSequenceClassification,
    pipeline
)
from thai2transformers.preprocess import process_transformers

# Load pre-trained tokenizer
tokenizer = CamembertTokenizer.from_pretrained(
                                  'airesearch/wangchanberta-base-att-spm-uncased',
                                  revision='main')
tokenizer.additional_special_tokens = ['<s>NOTUSED', '</s>NOTUSED', '<_>']

# Load pre-trained model
model = AutoModelForSequenceClassification.from_pretrained(
                                  'airesearch/wangchanberta-base-att-spm-uncased',
                                  revision='finetuned@wongnai_reviews')

classify_sequence = pipeline(task='sentiment-analysis',
          tokenizer=tokenizer,
          model=model)

dataset = load_dataset('wongnai_reviews', cache_dir='huggingface_cache')

nok = 0
for i in range(len(dataset['test'])):
      text = dataset['test'][i]['review_body']
      processed_input_text = process_transformers(text)
      processed_input_text_truncated = ''.join(tokenizer.tokenize(processed_input_text)[0:508]).replace('_', '')

      result = classify_sequence(processed_input_text_truncated)[0]
      y_pred = int(result['label'])
      y_true = int(dataset['test'][i]['star_rating'])
      nok += 1 if y_pred==y_true else 0

p = 100 * nok / len(dataset['test'])
print('Acc = %.3f' % p)
```

**Comment by artificiala** [comment]
สวัสดีค่ะ คุณ @peune,

ต้องขออภัยสำหรับการตอบที่ช้านะคะ 

สำหรับ accuracy ที่ไม่ตรงนี้ จะเกิดจาก บรรทัดต่อไปนี้นะคะ

```python
 processed_input_text_truncated = ''.join(tokenizer.tokenize(processed_input_text)[0:508]).replace('_', '')
```

โดยที่เป็นเพราะ symbol `_` (underscore, U+5F) จะต้องแก้ไขเป็น `▁` (U+2581)  แทนนะคะ สำหรับ symbol ▁ จะถูกใช้ใน SentencePiece เพื่อบ่งบอกว่ามี space อยู่ข้างหน้า subword หรือว่าเป็น subword แรกของประโยค ซึ่ง WangchanBERTa ที่ pre-trained มาจะไม่มีการใช้ symbol นี้ แต่ใช้ `<_>` แทนเพือแทนที่ space  การ replace symbol `_` (underscore) เลยทำให้ input format ผิดพลาดไปได้

```python
 processed_input_text_truncated = ''.join(tokenizer.tokenize(processed_input_text)[0:508]).replace('▁', '')
```

และบรรทัด
```python
y_pred = int(result['label'])
y_true = int(dataset['test'][i]['star_rating'])
```

เนื่องจากว่า star_rating ใน `wongani_reviews` ที่ host บน huggingface จะเริ่มจาก 0 - 4
จึงต้อง -1 ที่ y_pred เพราะว่า finetuned model จะ predict label เริ่มจาก 1-5 (จาก `model.config.id2label`)

```python
y_pred = int(result['label']) - 1
y_true = int(dataset['test'][i]['star_rating'])
```
